### ISSUE TITLE
WO-029: JAX port (gosplan/jax/)

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#34]
> ## NOT ISSUABLE YET - THIS CARD IS A PLACEHOLDER AND ITS CONTENT IS PROVISIONAL
>
> **This work order is issued only AFTER the Phase-2 spec revision.** It is gated on gate **G2** and
> on the lead's spec revision that follows G2. Nothing here may be started before both have happened.
>
> **The behaviour of the P2 sketch modules is frozen at that revision, and not before.** The port is
> a *translation* of `gosplan/env/`, and half of what it must translate - trade (PLAN §2.13), the
> ministry (§2.14), delivery timing, quality, targeted audits, soft budget - is a P2 sketch. Porting
> an unfrozen rule would produce two implementations of a rule that does not yet exist.
>
> **Everything under "Implementation notes" below is PROVISIONAL** until the lead revises the spec.
> Do not read it as settled and do not present it to anyone else as settled.

> **WO-029** - Phase 2 - Owner **LEAD - one unit, not delegated** (PLAN §1.3, finding F14) - Difficulty **set by the LEAD at issue** - Gate **G3**
> Card: [`workorders/WO-029.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-029.md)

## This card is LEAD work and is not delegated

PLAN §1.3, finding F14, names the step function, the JAX port and the PPO adapter as the three units
the **LEAD writes itself rather than delegating**. For this card that means, concretely: **no
implementer session writes into `gosplan/jax/`.** Delegating this card is itself listed under
Forbidden below. The issue exists so the port is visible in the DAG and so its parity criterion is on
the record; it is not an invitation for an implementer session to pick it up.

## What this is

A second implementation of the environment on a vectorised array backend, so the Phase-2 acceptance
sweep and the Phase-3 Sobol design are affordable. It is a *translation* of `gosplan/env/`, not a
redesign: the NumPy path stays the reference that the golden files are generated from, and the port
earns its keep only if it agrees with that reference exactly. The agreement is provable rather than
lucky because every random draw in the design is keyed rather than streamed, so the two backends
produce the same numbers without having to make the same calls in the same order.

## Depends on

- gate G2 - the Phase-1 gate (PLAN §4.5 criteria 1-4, signed off Human + LEAD). This card cannot be
  started before G2 has passed.
- the **P2 spec revision** (LEAD).
- WO-021
- WO-022
- WO-023
- WO-024
- WO-025

  (every Phase-2 mechanism must exist in NumPy before it can be ported and parity-checked)
- WO-009 - the NumPy step function that is the reference
- WO-004 - the keyed RNG, which is what makes parity meaningful
- WO-010 - `TruthfulMyopic`, the driver of the parity test

Consumed by WO-031 (the G3 condition "JAX parity") and by WO-032/WO-033/WO-034 - PLAN §14 makes
Sobol "JAX only" and budgets contrasts and the estimator-bias study on JAX.

**Blocked by:** #26 (WO-021), #27 (WO-022), #28 (WO-023), #29 (WO-024), #30 (WO-025), #14 (WO-009), #9 (WO-004), #15 (WO-010), #45 (gate G2)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- `PLAN.md` §§ 2.5-2.15, 10, 12.4, 13, 14 - PLAN.md is git-ignored and supplied out of band; it is
  not in the repository and has no blob link.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rules 4, 6, 8, 9
- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) - entire
- [`spec/CHANGELOG.md`](https://github.com/Xocas12/gosplan-env/blob/main/spec/CHANGELOG.md)
- [`gosplan/jax/__init__.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/jax/__init__.py) -
  the standing statement of what the port must satisfy
- [`gosplan/rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/rng.py)
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py)
- `gosplan/env/` - every module:
  [`state.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/state.py),
  [`production.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/production.py),
  [`planner.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/planner.py),
  [`reporting.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reporting.py),
  [`reward.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reward.py),
  [`obs.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/obs.py),
  [`prices.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/prices.py),
  [`trade.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/trade.py),
  [`ministry.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/ministry.py),
  [`step.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/step.py),
  [`env.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/env.py)
- [`gosplan/agents/heuristic.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/agents/heuristic.py) -
  `TruthfulMyopic`
- [`gosplan/metrics/ledger.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/metrics/ledger.py)
- [`ref/ref_step.py`](https://github.com/Xocas12/gosplan-env/blob/main/ref/ref_step.py)
- [`tests/golden/`](https://github.com/Xocas12/gosplan-env/blob/main/tests/golden)
- [`pyproject.toml`](https://github.com/Xocas12/gosplan-env/blob/main/pyproject.toml) - the `jax`
  optional-dependency group
- **the revised spec text, which does not exist yet - that is what gates this card**

Read nothing else. CONTRACT rule 12.

## Write only

Provisional; the revision fixes the final list.

- `gosplan/jax/**` - the port, LEAD-authored, one unit

Not this card: `gosplan/env/**` (the NumPy path stays the reference), `ref/`, `tests/unit`,
`tests/behavioural`, `tests/golden` (frozen, CONTRACT rule 2), `spec/spec.py`.

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

**No new test ids are frozen yet beyond the parity criterion below** - the revision writes the test
file and names its id in this section (CONTRACT rule 12).

The one acceptance criterion PLAN already states exactly (PLAN §12.4, and the G3 row of PLAN §13,
"JAX parity"):

- [ ] **100 agent-steps driven by `TruthfulMyopic` are identical to the NumPy implementation to
      `1e-5`, under the same key-based RNG**
- [ ] the **capital**, **IRS** and **setup-cost** branches are ported under `lax.cond` / `jnp.where`
      and **each gets its own parity case** - they are Phase-2 toggles, off in Phase 1, and therefore
      the easiest to get wrong without noticing
- [ ] the full frozen Phase-1 suite is green **on the NumPy path, unchanged**

Completion command, copy-paste:

```bash
uv run --extra jax ruff check . && uv run --extra jax pytest -q
```

The revision appends the parity-test invocation. `pytest -q` never collects `tests/acceptance/`
(CONTRACT rule 13).

## Implementation notes

**Provisional until the lead revises the spec.**

### Why the parity test is not a tautology

PLAN §2.15 keys every draw as

```
SeedSequence([seed_env, crc32(purpose), *indices])
```

so a draw is a pure function of `(seed_env, purpose, indices)` and is **order-independent**. The two
implementations therefore agree *by construction*, not because both happen to walk one stream in the
same order. A port that reproduced the trajectory only by preserving call order would be one refactor
away from silently diverging, and its green parity test would mean nothing.

### Branching

`jnp.where` where both sides are cheap and defined; `lax.cond` where evaluating the untaken side
would be wrong or wasteful. The discontinuities the design deliberately contains, and which the port
**must not round off**:

```
Λ_w(x) = 1[x ≥ 0] at w = 0, and the ρ_cap = ∞ branch of bonus      (§2.8)
the θ = ∞ (min) branch of the coverage aggregator                  (§2.6)
the claimed_i == 0 convention fill_i = 1 in deliver                (§2.7.3)
the σ_c → 1 Cobb–Douglas limit of the welfare index                (§2.9.3)
the setup cost F · 1[e > 0], a genuine discontinuity at e = 0      (§2.6)
```

A port that rounds a Heaviside into a sigmoid **has changed the economics, not the backend**:
`notch_width` is a treatment variable and the smooth arm is a configuration, so smoothing the notch
in the backend silently moves every arm onto the smooth treatment.

### The contract does not change with the backend

- `Array` is the alias every signature is written against (PLAN §10, `Array = np.ndarray`), and the
  port substitutes its own type behind that name. No module may rely on a numpy-only method in a
  signature.
- **CONTRACT rule 9** governs randomness identically: no global generator, no stream threading, and
  no `jax.random` key split standing in for a keyed draw.
- **CONTRACT rule 4** forbids a vectorised rollout any per-step shaping and any running reward
  normalisation. Vectorising is not a licence to normalise returns across the batch dimension.
- **CONTRACT rule 6** keeps whatever the port emits for the ledger off every agent-facing path.
- **CONTRACT rule 8** keeps the `at_bound` accounting intact, so `BOUND_BINDING` still fires at the
  same > 1% threshold on the ported path.

### Tolerance

**1e-5 over 100 agent-steps** is the criterion, and it is not negotiable downward. A port that needs
a looser tolerance has a bug, and the bug is the finding - it is reported as a defect, not as a
tolerance question.

### What the revision must settle before the port is written

The final Phase-2 dynamics themselves - in particular **the trade matcher of WO-024**, whose greedy
loop is the one piece of PLAN §§2.5-2.14 with no obvious vectorised form, and whose tie-breaking must
be **identical on both backends** for the parity test to mean anything.

CONTRACT rule 3: choosing "the reasonable default" for any unfrozen rule is a violation.

## Forbidden

- **Delegating this card.** PLAN §1.3 finding F14 keeps it with the LEAD.
- Changing `gosplan/env/` to make the port easier. The NumPy path is the reference for the golden
  files (T-B7); if a change to it is genuinely needed, it is a separate LEAD change with a
  `spec/CHANGELOG.md` entry and regenerated goldens (CONTRACT rules 1, 2).
- Smoothing any of the discontinuities listed above, "for gradient stability" or otherwise.
- Any `jax.random` call inside the ported environment (CONTRACT rule 9).
- Reporting a parity failure as a tolerance question. 1e-5 over 100 agent-steps is the criterion.
- Computing PLAN §4.1 rows **2, 5, 6, 7** - held out until WO-030.
- Editing `ref/`, `tests/unit`, `tests/behavioural`, `tests/golden` or `spec/spec.py`.

## Contract rules that bite here

- **Rule 3 - STOP AND REPORT.** Half the dynamics being ported are unfrozen sketches. Porting one by
  choosing "the reasonable default" is a violation; the port waits for the revision.
- **Rule 12 - work orders.** Read only the whitelist, write only `gosplan/jax/**`, run the completion
  command verbatim, report in the four-part format.
- **Rule 9 - RNG.** Keyed, order-independent draws only. No `jax.random` in the ported environment,
  no key-split standing in for a keyed draw.
- **Rule 4 - reward terms.** No shaping and no running reward normalisation in a vectorised rollout.
- **Rule 6 - welfare blindness.** Ledger output on the ported path stays off every agent-facing path.
- **Rule 8 - bounds are results.** `at_bound` accounting is preserved so `BOUND_BINDING` still fires.
- **Rules 1 and 2 - frozen spec and frozen tests.** The NumPy reference, `ref/` and the frozen test
  directories are not this card's.

## Definition of done

- [ ] 100 agent-steps with `TruthfulMyopic` identical to NumPy to 1e-5 under key-based RNG, with the
      measured maximum deviation reported
- [ ] Capital, IRS and setup-cost branches each have their own parity case
- [ ] The frozen Phase-1 suite green on the NumPy path, unchanged
- [ ] `uv run --extra jax ruff check . && uv run --extra jax pytest -q` clean, plus
      `uv run ruff format --check .`
- [ ] `uv run python scripts/contract_guard.py --all` clean
- [ ] No file outside `gosplan/jax/**` created, edited or deleted
- [ ] CI `ci-ok` green on the PR
- [ ] The four-part report posted as the PR body: 1) tests passed/failed with output, 2) ambiguity
      reports filed, 3) files changed, 4) anything you were tempted to invent and did not
